### ISSUE TITLE
Implement penover / pendown / penmove / penup events and similar for other new pointer types.

### ISSUE THREAD

**Issue by flackr** [w3c/pointerevents#359]
#203 proposes being able to assign different touch-action values by pointer type, however I think that the idea of telling the browser not to handle pen inputs, and adding a passive pointerevent listener to filter on pen events is a bit inefficient and also not as ergonomic as directly expressing what you want, which is to handle (and block) pen events.

I propose that we allow developers to instead add a pendown listener which will be called when PointerEvent events of type 'pointerdown' and pointerType 'pen' are generated. I think that this more directly expresses the developer intent to handle pen events, allows the browser to continue to accelerate touch scrolls, and is more efficient in that we don't even have to send the touch events to blink.

TLDR; the rough proposal here is when a pointer* event `e` of pointerType 'pen' occurs, we would:
First dispatch pointer* listeners with `e`.
Then dispatch pen* listeners with `e` (possibly only if e hadn't prevented).

e.g.
```js
elem.addEventListener('pendown', (evt) => {
  console.log('pendown');
});
elem.addEventListener('pointerdown', (evt) => {
  console.log('pointerdown');
});
```
Touching elem with the pointer would result in the following output:
```
pointerdown
pendown
```

This matches the order and roughly the [behavior with legacy events](https://jsfiddle.net/flackr/34Lvkrq8/6/) except that with legacy events we are not consistent about whether preventing the event in the pointer* listener prevents the dispatch of the type specific listener, and we would be invoking the listener with the generic PointerEvent event rather than adding a new PenEvent type.

**Comment by smaug----** [comment]
This is kind of the opposite what pointerevent have been trying to do. We have hoped to get rid of the legacy events - that is where the name 'legacy' comes from after all ;)

**Comment by flackr** [comment]
I totally understand, but we've lost the ability to know what type of events the developer wants to handle. The nice thing is that we wouldn't need to introduce new type specific events and developers could always expect to see them show up as pointer events - i.e. they don't need to add a new set of handlers each time a new input device type shows up.

**Comment by patrickhlauke** [comment]
yeah i'll admit that to me this would feel like a step back, philosophically - as we're trying to get devs to think in more device/input agnostic ways first and foremost. and if they *want* to handle events differently, it doesn't seem to onerous for them to switch/case when they get a pointer event to then have different behaviour at that point. also *if* new classes/types of pointer become popular in future, we'd then need to reinvent further companion events again (e.g. the mentioned eraser, or microsoft dial, as pointers ... it'd be trivial currently to just extend the type values to allow for these, but then magicking up eraserdown/eraserup/etc seems exactly the sort of thing we wanted to avoid by having a generic input-agnostic model)

**Comment by flackr** [comment]
I agree it's not hard for developers to have a switch on pointerType, the problem is that the browser can't know that they only care to handle a certain pointer type (e.g. pen for a drawing application) so the natural way of adding a listener like the following will slow down touch scrolling as well:

```js
canvas.addEventListener('pointerdown', (evt) => {
  if (evt.pointerType != 'pen')
    return;
  evt.preventDefault();
  // Start drawing.
});
```

Without input type specific handlers, the developer has to work around this by defining a pointer type specific touch action (in the linked issue) so that the events produce no default action and ensure that the element uses a passive event listener:

```html
<style>
canvas {
  pointer-action: pen(none);
}
</style>
<script>
canvas.addEventListener('pointerdown', (evt) => {
  if (evt.pointerType != 'pen')
    return;
  // Start drawing.
}, {passive: true});
</script>
```

However, it may be that the complexity of having additional event listener types is not worth it.

**Comment by NavidZ** [comment]
@flackr pointerevent listeners are passive by default. So even in your first example that listener is passive and that prevent default does not (should not) prevent scrolling. Although that is what I recall from all the discussions in the PEWG.
But I don't see it mentioned in the spec. Maybe I'm not searching for the right term.

**Comment by flackr** [comment]
Thanks @NavidZ, that certainly makes it less likely devs will accidentally added blocking listeners, though I still contend that blocking listeners is the most natural way for a developer to take over handling input on a certain element.

**Comment by flackr** [comment]
> yeah i'll admit that to me this would feel like a step back, philosophically - as we're trying to get devs to think in more device/input agnostic ways first and foremost. and if they _want_ to handle events differently,

Just to respond directly to this point. I don't think this is necessarily a step back as we would still require that all new events are also pointer events and so in most cases (and recommended practice) devs would handle them using the pointer* event handlers. It would only be in special cases where they wanted to handle a certain type differently that they may choose to add a specific type handler.

**Comment by smaug----** [comment]
@NavidZ where is it said that pointer event listeners would be passive?

**Comment by flackr** [comment]
@NavidZ Perhaps I misunderstood. It looks like even passing `{passive: false}` and calling `preventDefault()` on the events does not prevent anything? https://output.jsbin.com/dotowic

In that case it seems that the model for pointer events is that you could never use an event listener to prevent default behaviors which means we need a way of refining the pointer-action.

**Comment by NavidZ** [comment]
> @NavidZ where is it said that pointer event listeners would be passive?

You are right they are not passive. Perhaps what I'm remembering is not having a blocking default action (mainly around scrolling behavior as Chrome was very much concerned about scrolling performance) as outlined in this [table](https://w3c.github.io/pointerevents/#attributes-and-default-actions). 

> @NavidZ Perhaps I misunderstood. It looks like even passing {passive: false} and calling preventDefault() on the events does not prevent anything? https://output.jsbin.com/dotowic
>
>In that case it seems that the model for pointer events is that you could never use an event listener to prevent default behaviors which means we need a way of refining the pointer-action.

I see. This is probably confirming the above paragraph regarding default action. They don't have the scrolling interruptible default action as opposed to [touch events](https://w3c.github.io/touch-events/#cancelability).

**Comment by msalsbery** [comment]
> @NavidZ Perhaps I misunderstood. It looks like even passing `{passive: false}` and calling `preventDefault()` on the events does not prevent anything? https://output.jsbin.com/dotowic
> 
> In that case it seems that the model for pointer events is that you could never use an event listener to prevent default behaviors which means we need a way of refining the pointer-action.

I'm confused/lost here (but following, as I'm a heavy end user)...I've never had the impression that preventDefault did nothing in pointer event handlers. In that linked example, for example, the preventDefault call prevents text selection with a mouse device (on Edge Chromium at least). That's the type of behavior I've always expected...has that changed?

**Comment by NavidZ** [comment]
> In that linked example, for example, the preventDefault call prevents text selection with a mouse device (on Edge Chromium at least). That's the type of behavior I've always expected...has that changed?

I did retract my comment in the [latest comment](https://github.com/w3c/pointerevents/issues/359#issuecomment-823274011).
To reiterate and be more accurate they are not passive (in the way that DOM spec says) as you can prevent default a pointerdown and it will cause the corresponding mouse events not to fire. But the default action of pointerevents has nothing to do with scrolling as already [specified](https://w3c.github.io/pointerevents/#attributes-and-default-actions) in the PointerEvent spec. 

Regarding the text selection I believe that is the default action of mouse events (and by extension pointerevents that generate those mouseevents) which are either specific behavior of Blink (or other engines) or if it is spec'ed it should be in uievents spec.

**Comment by flackr** [comment]
FWIW, I think it's confusing that the [touch-action property](https://w3c.github.io/pointerevents/#details-of-touch-action-values) specified in pointer-events "determines whether touch input MAY trigger the user agent's **default behavior**" which includes scrolling, but that default behavior is not to be confused with the [default action](https://w3c.github.io/pointerevents/#attributes-and-default-actions). I would find having a note in the default actions table pointing out that we intentionally do not consider panning / scrolling a default action, despite being a default behavior very helpful.

**Comment by patrickhlauke** [comment]
> FWIW, I think it's confusing that the [touch-action property](https://w3c.github.io/pointerevents/#details-of-touch-action-values) specified in pointer-events "determines whether touch input MAY trigger the user agent's **default behavior**" which includes scrolling, but that default behavior is not to be confused with the [default action](https://w3c.github.io/pointerevents/#attributes-and-default-actions). I would find having a note in the default actions table pointing out that we intentionally do not consider panning / scrolling a default action, despite being a default behavior very helpful.

incidentally, this confusion between default *behavior* and default *action* is something I mentioned in a recent call - see https://github.com/w3c/pointerevents/pull/349#issuecomment-811092269

definitely something we want to reword/address (for expediency, we decided in the call to split out this aspect so we could merge #349, but i'll be circling back to this)

edit: split this out into a fresh separate issue https://github.com/w3c/pointerevents/issues/368

**Comment by flackr** [comment]
Shall we close this issue? It seems the model with pointer events is that you can never prevent panning / zooming, so adding pen specific pointerevent listeners would not address the issue without reversing the default action not being panning / zooming. We do in fact need to add pointer-type specific touch-actions to control panning / zooming by pointer type.

**Comment by patrickhlauke** [comment]
Sure, let's do it.
